### PR TITLE
Add a Mime Type for .exe and .dll files.

### PR DESCRIPTION
**Pull request by KirillOsenkov**
Useful for serving downloads of installers.
Could be application/octet-stream, but I hear application/vnd.microsoft.portable-executable is specified by IANA here:

https://www.iana.org/assignments/media-types/application/vnd.microsoft.portable-executable

### CODE REVIEW

**Comment by shirhatti** [comment]
https://github.com/dotnet/aspnetcore/blob/595f11b6c87547d5a9c9c6d626128fef6f51e356/src/Middleware/StaticFiles/src/FileExtensionContentTypeProvider.cs#L14-L17

**Comment by KirillOsenkov** [comment]
I'm guessing a lot of people publish installers, downloaders, setup.exe etc. It is a bit counter-intuitive that this doesn't work out of the box.

**Comment by Tratcher** [comment]
Historically IIS didn't include these because the static files directory and your app directory were the same, it didn't want to expose your app assemblies as static files. That's less of an issue in AspNetCore since we separate out static content to a subdirectory in the default layout.

> Useful for serving downloads of installers.

I buy that for exes, but what about dlls?

**Comment by KirillOsenkov** [comment]
Yes, I can image .dll download is not such a common scenario.

**Comment by Tratcher** [comment]
@shirhatti we should consider adding exe's as they are quite common.

**Comment by KirillOsenkov** [comment]
For my own curiosity, what are the concerns with adding mime types to this list? From my naive perspective this should be pretty much benign?

**Comment by Tratcher** [comment]
> For my own curiosity, what are the concerns with adding mime types to this list? From my naive perspective this should be pretty much benign?

The list is already enormous and likely causing some performance issues. We also get very frequent requests to add new entries so we tried to establish clear criteria.

**Comment by KirillOsenkov** [comment]
Performance issues should be an orthogonal concern from delivering a good default experience to our users.

If the performance is inadequate, let's prove it, let's measure it and let's fix it. I don't believe having more extensions in that dictionary should affect performance in any way if we do it right.

This raises questions: why is this dictionary created as an instance every time we create the class? Why isn't there a shared static Dictionary somewhere that this falls back to?

I honestly would optimize for the principle of least surprise: as a user, I just want to serve some static files. I shouldn't be thinking about Mime types, IIS, or some legacy reasons. My out of the box experience is that I serve a file, of any IANA supported extension, and it just magically works out of the box, no assembly required.

Thanks!

**Comment by Tratcher** [comment]
Fair, the performance comment was lazy on my part, I don't have the data handy to prove that.

> why is this dictionary created as an instance every time we create the class

There's commonly only one instance of this class an application anyways.

Let me clarify the other comment. We get very frequent requests for additions to this list and many of them are for esoteric extensions that do not have broad usage or well defined standards. Adding such content provides little value, we'd rather those developers edit their own list locally.

**Comment by KirillOsenkov** [comment]
OK understood.

File: src/Middleware/StaticFiles/src/FileExtensionContentTypeProvider.cs [hunk 1]
```diff
@@ -84,6 +84,7 @@ public FileExtensionContentTypeProvider()
                 { ".dib", "image/bmp" },
                 { ".dir", "application/x-director" },
                 { ".disco", "text/xml" },
+                { ".dll", "application/vnd.microsoft.portable-executable" },
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/src/FileExtensionContentTypeProvider.cs:87]:
```suggestion
```

File: src/Middleware/StaticFiles/src/FileExtensionContentTypeProvider.cs [hunk 2]
```diff
@@ -104,6 +105,7 @@ public FileExtensionContentTypeProvider()
                 { ".eps", "application/postscript" },
                 { ".etx", "text/x-setext" },
                 { ".evy", "application/envoy" },
+                { ".exe", "application/vnd.microsoft.portable-executable" },
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/src/FileExtensionContentTypeProvider.cs:108]:
```suggestion
                { ".exe", "application/vnd.microsoft.portable-executable" }, // https://www.iana.org/assignments/media-types/application/vnd.microsoft.portable-executable
```

**Review by Tratcher — COMMENTED** [review]
Triage: We should add exe support, but we don't plan on adding dll support.

I do want to try this and see how browsers handle that "application/vnd.microsoft.portable-executable" content-type.

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**